src/PDO/Peachpie.Library.PDO/PDOStatement.cs

```diff
@@ -485,40 +485,63 @@ public virtual PhpArray fetchAll(PDO_FETCH fetch_style = default, PhpValue fetch
 
 
             // Handle FETCH_GROUP case 
-            if (flags == PDO_FETCH.FETCH_GROUP && !result.IsEmpty())
+            if (flags == PDO_FETCH.FETCH_GROUP)
             {
-                // hasNumKeys: when FETCH_NUM or FETCH_BOTH
-                var hasNumKeys = result[0].ToArray().ContainsKey((IntStringKey)0);
+                result = GroupResult(result, style);
+            }
 
-                var groupedResult = new PhpArray();
-                foreach (var kp in result)
-                {
-                    var row = kp.Value.ToArray();
+            //
+            return result;
+        }
 
-                    // We remove the first column and use it to group rows
-                    var firstCol = row.RemoveFirst().Value;
+        /// <summary>
+        /// Create result set grouped by the forst column according to <see cref="PDO.FETCH_GROUP"/>.
+        /// </summary>
+        private protected static PhpArray/*!!*/GroupResult(PhpArray result, PDO_FETCH style)
+        {
+            if (result.IsEmpty())
+            {
+                return result;
+            }
 
-                    // For numeric keys
-                    // => Always remove the 0 key (FETCH_NUM: does nothing, FETCH_BOTH: remove remaining)
-                    // => Reindex starting from 0
-                    if (hasNumKeys)
-                    {
-                        row.RemoveKey((IntStringKey)0);
-                        row.ReindexIntegers(0);
-                    }
+            // hasNumKeys: when FETCH_NUM or FETCH_BOTH
+            var hasNumKeys =
+                (style & ~PDO_FETCH.Flags) == PDO_FETCH.FETCH_NUM ||
```

Comment:
Beware for hasNumKeys, you need to also check for `FETCH_DEFAULT` since FETCH_BOTH is the default...

It was partly why I choose to go to the "check if there is a numeric key in the first row" route...

Reply:
good point!